### DISCUSSION TITLE
groups of comparisons

### DISCUSSION THREAD

**Discussion by msiemionCalistapw** [moj-analytical-services/splink#730] [Q&A] [answered]
Hello,

Is there a way to program groups of comparisons into settings using and/or logic?

For example, if I wanted to compare (first name AND street name) OR (first name AND date of birth). 

Thank you!

**Comment by ThomasHepworth** [comment, marked as answer]
This is something we have been meaning to add more info on in the demos for a few weeks, but haven't found the time.

Anyway, to generate specific comparisons like you've mentioned above, you simply need to tweak the blocking rules within your settings object.

For example, if you wanted the rules you've proposed above, you'd want to use a settings object that looks like so:
```python
settings = {
    "probability_two_random_records_match": 1/100_000,
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.surname = r.surname AND l.first_name = r.first_name",
        "l.first_name = r.first_name AND l.dob = r.dob" # multiple rules are equivalent to OR clauses
    ],
    "comparisons": [
        # enter your comparisons here
    ]
}
```

The key is that you just need to specify the `l.` and `r.` prefix for each column you want to evaluate.

<hr>

If you want some more adventurous comparisons, such as blocking on similarity functions (levenshtein, jaro-winkler, etc), please see [this discussion](https://github.com/moj-analytical-services/splink/discussions/461).

You can also now profile your blocking rules, to see how many comparisons they are generating before you run any predictions. Some demo code for this is available [here](https://gist.github.com/ThomasHepworth/26d5cc2fee202cd6738def388e166e19).

**Reply by msiemionCalistapw** [reply]
great, thank you! Would comparisons need to be written as such: 
          exact_match("surname ")  AND exact_match("first_name"), 
          exact_match("first_name")  AND exact_match("dob"), 
???

**Reply by RobinL** [reply]
Please see Method 3 [here](https://moj-analytical-services.github.io/splink/topic_guides/customising_comparisons.html#method-3-providing-the-spec-as-a-dictionary) 

As an example, the `sql_condition` would be `"surname_l = surname_r and first_name_l = first_name_r"`

To clarify:
- The blocking rules determine which pairwise record comparisons are generated
- The `Comparisons` then concern how to score those comparisons
